server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -259,44 +360,92 @@ public void onTaskCompleted(Task task) {
             return;
         }
 
-        if (task instanceof SearchShardTask == false) {
+        if (task instanceof SearchBackpressureTask == false) {
             return;
         }
 
-        SearchShardTask searchShardTask = (SearchShardTask) task;
-        if (searchShardTask.isCancelled() == false) {
-            state.incrementCompletionCount();
+        CancellableTask cancellableTask = (CancellableTask) task;
+        boolean isSearchTask = task instanceof SearchTask;
```

Comment:
Is it possible to add a generic class like `SearchBackpressureCancellableTaskPicker<T>` where `T` can be either of `SearchTask` or `SearchShardTask? With this, I believe the if check would be only once to invoke corresponding class and after that corresponding class would handle only one type. Thoughts?

Reply:
As discussed, adding a method to return the task type, given the task object. This will reduce the number of `instanceof` calls and code branches.